cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_csr_impl.hpp

```diff
@@ -141,77 +137,52 @@ sycl::event assign_clusters(sycl::queue& q,
                             pr::ndview<Float, 2>& closest_dists,
                             const event_vector& deps = {}) {
     ONEDAL_PROFILER_TASK(assign_clusters, q);
-    auto data_squares_ptr = data_squares.get_data();
-    auto cent_squares_ptr = centroid_squares.get_data();
-    auto responses_ptr = responses.get_mutable_data();
-    auto closest_dists_ptr = closest_dists.get_mutable_data();
-    // Calculate rest part of distances
-    auto dist_event = custom_spgemm(q,
-                                    values,
-                                    column_indices,
-                                    row_offsets,
-                                    centroids,
-                                    distances,
-                                    Float(-2.0),
-                                    Float(0),
-                                    deps);
 
-    const auto distances_ptr = distances.get_data();
+    // Workaround. Sparse gemm cannot accept transposed dense inputs in oneMKL 2025.0.
+    // Error text:
+    // oneapi::mkl::sparse::gemm: unimplemented functionality: Only non-transpose
+    // operation is supported for dense matrix
+    // TODO: Remove separate transpore and pass centroids.t() into gemm after updating
+    //       to oneMKL that supports transposed dense input matrix.
+    auto centroids_transposed =
+        pr::ndarray<Float, 2>::empty(q,
+                                     { centroids.get_dimension(1), centroids.get_dimension(0) },
+                                     sycl::usm::alloc::device);
 
-    const auto cluster_count = centroids.get_dimension(0);
-    const auto row_count = static_cast<size_t>(row_offsets.get_count() - 1);
-    // based on bechmarks an optimal block size is equal to 8 work-group sizes
-    const std::int64_t block_multiplier = 8;
-    const std::int64_t row_block = block_multiplier * bk::device_max_wg_size(q);
+    sycl::event transpose_event = transpose(q, centroids, centroids_transposed, deps);
```

Comment:
Can it be replaced with .t() method of ndarray?

Reply:
No. Because t() method does not do the actual data transpose, it just changes ndarray's layout from row_major to column_major (or vise versa). It is not enough to call the correct variant of sparse gemm.